### DISCUSSION TITLE
Publishing directly to Docker Hub

### DISCUSSION THREAD

**Discussion by pfedotovsky** [dotnet/sdk-container-builds#359] [General]
Hi,

I want to share my experience with publishing directly to the Docker Hub. I hope it'll be helpful for someone and used to update the documentation: https://github.com/dotnet/sdk-container-builds/blob/main/docs/RegistryAuthentication.md.

I'm using macOS Ventura with an Apple M1 chip.
First, I've found that they're at least three Docker Hub registry URLs which is confusing:

1. **docker.io**: seems to be used by default in docker on macOS
2. **registry-1.docker.io**: mentioned in https://github.com/dotnet/sdk-container-builds/blob/main/docs/RegistryAuthentication.md
3. **registry.hub.docker.com**: used when publishing from Visual Studio on Windows

To authenticate with registries, I used `docker login <registry-name>,` i.e., `docker login registry.hub.docker.com`
To publish, I used the command: 
`dotnet publish --os linux --arch x64 -c Release /p:ContainerRegistry=<registry-name> /p:ContainerImageName=<dockerhub-username>/<repository-name>`
Example:
`dotnet publish --os linux --arch x64 -c Release /p:ContainerRegistry=registry.hub.docker.com /p:ContainerImageName=pfedotovsky/dotnetru-appservice`

Now, the _registry.hub.docker.com_ and _registry-1.docker.io_ are working correctly, while with _docker.io_, the _dotnet publish_ command executes successfully, but no package is found on Docker Hub.

**Comment by baronfel** [comment]
Thanks for this feedback @pfedotovsky - let me make sure I understand the takeaway here.

* We should mention all three Docker URIs that look like registries to users in our docs
* We should clarify that either of the `registry-1.docker.io` or `registry.hub.docker.com` registries should be used when logging in in our docs
* (Bonus) we should add some registry-specific checks (like have been done for Amazon Elastic Container Registry endpoints) that issue a warning
  * if the `docker-io` URI is used as the `ContainerRegistry` (because it won't work), or 
  * the `registry*` Docker URIs are used as the `ContainerRegistry` but we only have credentials for `docker.io` (because the user used a 'normal' `docker login` command)

Is there anything else you'd like to see?

**Reply by pfedotovsky** [reply]
@baronfel, thank you for the summary! Looks great, I don't have anything to add 😊
